x-pack/platform/plugins/shared/onechat/server/services/tools/builtin/definitions/create_visualization/graph_maps.ts

```diff
@@ -0,0 +1,617 @@
+  error: Annotation<string | null>(),
+});
+
+type MapState = typeof MapStateAnnotation.State;
+
+export const createMapGraph = (
+  model: ScopedModel,
+  logger: Logger,
+  esClient: IScopedClusterClient,
+  dataViewsService: DataViewsService
+) => {
+  // Node: Explore data views on demand
+  const exploreDataViewsNode = async (state: MapState) => {
+    logger.debug('Exploring data views for map');
+
+    // Get the last request_dataviews action to know what to explore
+    const lastRequestAction = [...state.actions].reverse().find(isRequestDataViewsAction);
+
+    if (!lastRequestAction) {
+      const action: ExploreDataViewsAction = {
+        type: 'explore_dataviews',
+        success: false,
+        requestId: 'unknown',
+        description: 'No request found',
+        error: 'No data view exploration request found in actions',
+      };
+      return {
+        actions: [action],
+      };
+    }
+
+    let action: ExploreDataViewsAction;
+    try {
+      // Get all data views
+      let dataViews = await dataViewsService.getIdsWithTitle();
+
+      // Filter by pattern if provided
+      if (lastRequestAction.pattern) {
+        const regexPattern = new RegExp(lastRequestAction.pattern.replace(/\*/g, '.*'));
+        dataViews = dataViews.filter((dv) => regexPattern.test(dv.title));
+      }
+
+      // If no data views, return empty
+      if (dataViews.length === 0) {
+        action = {
+          type: 'explore_dataviews',
+          success: false,
+          requestId: lastRequestAction.requestId,
+          description: lastRequestAction.description,
+          error: lastRequestAction.pattern
+            ? `No data views found matching pattern: ${lastRequestAction.pattern}`
+            : 'No data views found',
+        };
+      } else {
+        // Use LLM to select most relevant data views
+        const limit = lastRequestAction.limit || 5;
+        const prompt = `Given the following Kibana data views, select the ${limit} most relevant one(s) for this query: "${
+          lastRequestAction.description
+        }"
+
+Available data views:
+${dataViews.map((dv) => `- ID: ${dv.id}, Title: ${dv.title}`).join('\n')}
+
+Return ONLY a JSON array of objects with the selected data view IDs and a brief reason why each was selected.
+Format: [{ "id": "data-view-id", "reason": "brief explanation" }]
+
+Select up to ${limit} data view(s).`;
+
+        try {
+          const response = await model.chatModel.invoke([['user', prompt]]);
+          const content =
+            typeof response.content === 'string'
+              ? response.content
+              : JSON.stringify(response.content);
+
+          // Extract JSON from markdown code blocks or plain text
+          const jsonMatch =
+            content.match(/```(?:json)?\s*(\[[\s\S]*?\])\s*```/) || content.match(/(\[[\s\S]*?\])/);
+
+          if (!jsonMatch) {
+            throw new Error('Failed to parse LLM response');
+          }
```

Comment:
Is that a duplication of the `dataview_explorer` implementation (`x-pack/platform/plugins/shared/onechat/server/services/tools/builtin/definitions/dataview_explorer.ts`)? If so, we need to factorize (this is why I shove everything in that `onechat-genai-utils` package, fwiw, even if I'm not saying the best approach)